support-frontend/assets/components/subscriptionCheckouts/address/addressFields.tsx

```diff
@@ -75,7 +82,97 @@ function statesForCountry(country: Option<IsoCountry>): React.ReactNode {
 	}
 }
 
+type ValidityStateError = 'valueMissing' | 'patternMismatch';
+
 export function AddressFields({ scope, ...props }: PropTypes): JSX.Element {
+	const patternMismatch = 'Please use only letters, numbers and punctuation.';
+	const errorMessages: Record<
+		string,
+		{
+			[key in ValidityStateError]?: string;
+		}
+	> = {
+		lineOne: {
+			valueMissing: `Please enter a ${scope} address.`,
+			patternMismatch,
+		},
+		lineTwo: {
+			patternMismatch,
+		},
+		city: {
+			valueMissing: `Please enter a ${scope} town/city.`,
+			patternMismatch,
+		},
+		state: {
+			valueMissing:
+				props.country === 'CA'
+					? `Please select a ${scope} province/territory.`
+					: `Please select a ${scope} state.`,
+			patternMismatch,
+		},
+		postCode: {
+			valueMissing: `Please enter a ${scope} ${
+				props.country === 'US' ? 'ZIP code' : 'postcode'
+			}`,
+			patternMismatch,
+		},
+	};
+	const getErrorMessage = (
+		field: keyof AddressFieldsType,
+		validityStateError: ValidityStateError,
+	): string => {
+		return errorMessages[field][validityStateError] ?? '';
+	};
+	const setErrorsOnInvalid = (
+		event:
+			| React.FormEvent<HTMLInputElement>
+			| React.FormEvent<HTMLSelectElement>,
+		field: keyof AddressFieldsType,
+	) => {
+		/**
+		 * We only pass/use setErrors on the generic checkout
+		 * If it's not been passed to AddressFields then Redux handles
+		 * validation errors.
+		 **/
+		if (!props.setErrors) {
+			return;
+		}
+
+		preventDefaultValidityMessage(event.currentTarget);
+		const validityState = event.currentTarget.validity;
+		if (validityState.valid) {
+			// remove all errors for the field
+			props.setErrors(
+				props.errors.filter((error) => {
+					return error.field != field;
+				}),
+			);
+		} else {
+			const possibleValidityStateErrors: ValidityStateError[] = [
+				'valueMissing',
+				'patternMismatch',
+			];
+			const updatedErrors: AddressFormFieldError[] = [
+				// remove all errors for field
+				...props.errors.filter((error) => {
+					return error.field != field;
+				}),
+				// add all unresolved errors for the field
+				...(possibleValidityStateErrors
+					.map((possibleValidityStateError) => {
+						if (validityState[possibleValidityStateError]) {
+							return {
+								field,
+								message: getErrorMessage(field, possibleValidityStateError),
+							};
+						}
+					})
+					.filter(Boolean) as AddressFormFieldError[]),
```

Comment:
What are we filtering out here? Is it empty strings or null/undefined values?

Reply:
Could we use https://github.com/guardian/csnx/tree/main/libs/@guardian/libs/src/isNonNullable and get rid of the `as` ?

Reply:
> What are we filtering out here? Is it empty strings or null/undefined values?

Yes, the `.map` which iterates over `const possibleValidityStateErrors: ValidityStateError[] = ['valueMissing','patternMismatch']` can return an array with `undefined` items in it, if the `validityState` doesn't contain a corresponding error. I'm removing all the `undefined` items from that returned array here.

I wasn't happy having to define `possibleValidityStateErrors` tbh, the first thing I attempted was to iterate over `validityState`, which looks like an Object with keys, like below, unfortunately I couldn't iterate over this object, so I introduced `possibleValidityStateErrors` which I could then use to interrogate `validityState`.

<img width="642" alt="Screenshot 2024-08-27 at 12 30 30" src="https://github.com/user-attachments/assets/07071b08-da13-40c8-8b66-e3274edfeebb">

Reply:
> Could we use https://github.com/guardian/csnx/tree/main/libs/@guardian/libs/src/isNonNullable and get rid of the `as` ?

:eyes I'll take a look at this